linker: warn about orphan sections (again)

dcpleung opened the pull request:
(Previous patch set was reverted due to issue with priv_stack.
 Resubmitting after fixing the faults caused by priv_stack.noinit
 not at the end of RAM.)

This adds a linker flag and necessary changes to linker scripts
so that linker will warn about orphan sections.

Relates to #5534.

Fixes #10473, #10474, #10515.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Tested with sam_e70_xplained, nrf52_pca10040 and frdm_k64f. Sanitycheck passed except those mentioned in #10633, #10634 and #10508 (where these fail on master).

aescolar reviewed (dismissed):
It seems you have left behind the fixes related to https://github.com/zephyrproject-rtos/zephyr/issues/10493

dcpleung commented:
@aescolar Oops... I have updated the patch set with the exclusion of posix.

aescolar commented:
@dcpleung : Would it make sense to have this behind a Kconfig option, even if that option defaulted to y?I'm thinking that, if somebody has trouble with this feature, that would allow them to disable it easily.

dcpleung committed:
linker: allow SoC to insert linker script fragments

This allows the SoC to specify some additional linker script
fragments into the bss, data and read-only data sections.

For example, the Cypress PSOC6 has a few input sections that
must be put into bss and data sections. Without specifying
these in the linker script, they are consider orphan sections
and the placement is based on linker heuristic which is
arbitrary.

POSIX is not supported as the main linker script is
provided by the host system's binutils and we have no control
over it. Also, currently Xtensa SoCs have their own linker
scripts so there is no need to this feature.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
linker: warn about orphan sections

(Previous patch set was reverted due to issue with priv_stack.
 Resubmitting after fixing the faults caused by priv_stack.noinit
 not at the end of RAM.)

This adds a linker flag and necessary changes to linker scripts
so that linker will warn about orphan sections.

Relates to #5534.

Fixes #10473, #10474, #10515.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
arm: cypress/psoc6: add SoC specific linker input sections

The Cypress PSoC6 specifies some input sections in the startup
scripts. These sections (.heap, .stack, etc.) need to be placed
at correct location.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

AdithyaBaglody committed:
linker: arm: place priv.noinit at the end of RAM

This puts the priviledged stack at the end of RAM.
This combines PR #10507 and #10542.

Fixes #10473
Fixes #10474
Fixes #10515

Signed-off-by: Adithya Baglody <adithya.nagaraj.baglody@intel.com>
Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Added a kconfig to set different behavior for orphan section handling.
